census/keras/trainer/task.py

```diff
@@ -106,25 +111,34 @@ def dispatch(train_files,
       write_graph=True,
       embeddings_freq=0)
 
-  #TODO: This needs to be fixed in h5py so that writes to GCS are possible
-  # Don't attempt to create checkpoints on Cloud ML Engine for now because
-  # h5py doesn't come with native GCS write capability
-  if job_dir.startswith('gs://'):
-    callbacks=[evaluation, tblog]
-  else:
-    callbacks=[checkpoint, evaluation, tblog]
+  callbacks=[checkpoint, evaluation, tblog]
 
   census_model.fit_generator(
       model.generator_input(train_files, chunk_size=CHUNK_SIZE),
       steps_per_epoch=train_steps,
       epochs=num_epochs,
       callbacks=callbacks)
 
-  census_model.save(os.path.join(job_dir, CENSUS_MODEL))
+  census_model.save(path_join_for_non_gcs(job_dir, CENSUS_MODEL))
+  maybe_copy_file_to_gcs(job_dir, CENSUS_MODEL)
 
   # Convert the Keras model to TensorFlow SavedModel
   model.to_savedmodel(census_model, os.path.join(job_dir, 'export'))
 
+# Unhappy hack to work around h5py not being able to write to GCS.
+# Force snapshots and saves to local filesystem, then copy them over to GCS.
+def maybe_copy_file_to_gcs(job_dir, file_path):
+  if job_dir.startswith("gs://"):
+    with file_io.FileIO(file_path, mode='r') as input_f:
+      with file_io.FileIO(os.path.join(job_dir, file_path), mode='w+') as output_f:
+          output_f.write(input_f.read())
+
+def path_join_for_non_gcs(job_dir, file_path):
```

Comment:
why do you need this function?

Reply:
I think there is a problem in the way this function exists (in isolation) and gives the impression of its not so usefulness. I now see whats going on here. You want to separate how the write happens based on if the job_dir is gcs. You should explicitly do that (making the semantics clear). The current way is overloading the semantics into this function which doesn't read well.

Reply:
OK, what would you propose? Inline this `if` statement at the call sites?

Reply:
Yes I would just put this inline (and add one line comment as well on those if conditions).

Reply:
ok, I put all the calls inline. it's a bit more verbose now, not sure for the better, but it does make the logic more explicit.